File: platforms/ios/Tcall/ViewModels/AuthViewModel.swift
import Foundation

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var user: UserDto?
    @Published var loading = true
    @Published var submitting = false
    @Published var error: String?

    func restore() async {
        loading = true
        defer { loading = false }
        user = try? await ApiClient.shared.restoreSession()
    }

    func login(email: String, password: String) async {
        submitting = true
        error = nil
        defer { submitting = false }
        do {
            user = try await ApiClient.shared.login(email: email, password: password)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func logout() async {
        await ApiClient.shared.logout()
        user = nil
    }
}

File: android/app/src/main/kotlin/uz/vizara/tcall/ui/auth/AuthViewModel.kt
package uz.vizara.tcall.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import uz.vizara.tcall.data.AuthRepository
import uz.vizara.tcall.network.UserDto

data class AuthUiState(
    val loading: Boolean = true,
    val submitting: Boolean = false,
    val user: UserDto? = null,
    val error: String? = null,
)

class AuthViewModel(
    private val authRepository: AuthRepository,
) : ViewModel() {
    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    init {
        viewModelScope.launch {
            val user = authRepository.restoreSession()
            _state.value = AuthUiState(loading = false, user = user)
        }
    }

    fun login(email: String, password: String) {
        if (email.isBlank() || password.isBlank()) {
            _state.value = _state.value.copy(error = "Email va parolni kiriting")
            return
        }
        viewModelScope.launch {
            _state.value = _state.value.copy(submitting = true, error = null)
            authRepository.login(email, password)
                .onSuccess { user ->
                    _state.value = AuthUiState(loading = false, user = user)
                }
                .onFailure { err ->
                    _state.value = _state.value.copy(
                        submitting = false,
                        error = err.message ?: "Xatolik",
                    )
                }
        }
    }

    fun logout() {
        viewModelScope.launch {
            authRepository.logout()
            _state.value = AuthUiState(loading = false, user = null)
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }
}
